channeldb/db.go

```diff
@@ -141,11 +142,43 @@ var (
 // schedules, and reputation data.
 type DB struct {
 	kvdb.Backend
+
 	dbPath string
 	graph  *ChannelGraph
 	clock  clock.Clock
 }
 
+// Update is a wrapper around walletdb.Update which calls into the extended
+// backend when available. This call is needed to be able to cast DB to
+// ExtendedBackend.
+func (db *DB) Update(f func(tx walletdb.ReadWriteTx) error) error {
+	if v, ok := db.Backend.(kvdb.ExtendedBackend); ok {
+		return v.Update(f)
+	}
+	return walletdb.Update(db, f)
+}
+
+// View is a wrapper around walletdb.View which calls into the extended
+// backend when available. This call is needed to be able to cast DB to
+// ExtendedBackend.
+func (db *DB) View(f func(tx walletdb.ReadTx) error) error {
```

Comment:
Is this layer also needed? Since all callers end up using `kvdb.Update/View`? 

Guess it gives callers a way to just interact directly w/ the DB itself.

Reply:
Also in certain areas, there's a subtlety w.r.t what actually gets passed into `kvdb.Update`. In that behavior can differ if it's this struct (that embeds) a backend vs passing in the raw backend value.

Reply:
This is needed to keep the current global `kvdb.Update/View` functions while also allowing extended transaction logic when a backend implements that. With the `kvdb.ExtendedBackend` we keep proper separation, so no need to bake in reference to etcd here.

Reply:
> Also in certain areas, there's a subtlety w.r.t what actually gets passed into `kvdb.Update`. In that behavior can differ if it's this struct (that embeds) a backend vs passing in the raw backend value.

In my tests when `channeldb.DB` itself did not implement the `ExtendedBackend` interface, it was impossible to to the type assertions as with the embedding the extended type information is lost and `db.Backend` cannot be "upcasted" anymore.